[KEP-3203] Fetch and Render CVE JSON Feed

PushkarJ opened the pull request:
## Description

- Pull JSON blob from queried issues
- Use layout output formats + templates to generate HTML table and JSON blob
- Add localized strings and caption for CVE feed
- Add a new page to describe details about CVE feed and how to use it
- Update existing pages and link the official CVE feed from it

## Notes for Reviewers

### Why are we regenerating JSON blob from the JSON blob in the bucket ?
- Not all fields stored in the bucket are needed 
- JSON download directly to a file with Hugo data templates is not possible
- Gives us better control and ability to add extra fields as needed
- Pulling data through hugo data templates allow us to reuse built-in caching mechanisms

### Why not version control the JSON blob?

To avoid multiple sources of truth. The content in the bucket will remain the only source of truth as the JSON blob is constructed from GitHub issues with label `official-cve-feed` (See this for more info: https://github.com/kubernetes/test-infra/pull/23428). However, as a break glass, there is an option to temporarily push the JSON blob by overriding the layouts json file with raw JSON in cases where push to bucket fails or if bucket is inaccessible. 

### Are the locations for the layouts files correct?

Current locations are mostly my guess, so happy to refactor them as needed

TODO:

- [x] Update JSON blob location to k8s GCS bucket
- [x] Add more description in the PR
- [x] Address PR feedback from superseded PR

Preview: https://deploy-preview-35228--kubernetes-io-main-staging.netlify.app/docs/reference/issues-security/official-cve-feed/
KEP tracker: https://github.com/kubernetes/enhancements/issues/3203
xref: https://github.com/kubernetes/sig-security/issues/1
supersedes https://github.com/kubernetes/website/pull/31051 and https://github.com/kubernetes/website/pull/34765
/hold
/sig security docs

sftim commented on layouts/shortcodes/issues-security.html:
```diff
@@ -0,0 +1,19 @@
+<table class="security-cves">
+  <caption style="text-align:center">Table: Official Kubernetes CVE List</caption>
```
This should be a localized string, really.

sftim replied:
Oh, also try to avoid putting CSS here. Instead, set a style rule that uses a `table.security-cves caption` selector.

PushkarJ replied:
Updated

sftim reviewed (commented):
- Paths within `static` that are actually generated should include `generated` in their path. A `find` and `grep` should show some existing paths that follow that convention.
- I'd name `gen-cve-feed.sh` as `fetch-cve-feed.sh`

:question: How do we make sure that people don't accidentally commit that fetched file?

:question: For a container build, shouldn't the fetch happen _inside_ the container?

sftim commented on Makefile:
```diff
@@ -19,6 +19,9 @@ CCEND=\033[0m
 help: ## Show this help.
 	@awk 'BEGIN {FS = ":.*?## "} /^[a-zA-Z_-]+:.*?## / {sub("\\\\n",sprintf("\n%22c"," "), $$2);printf "\033[36m%-20s\033[0m %s\n", $$1, $$2}' $(MAKEFILE_LIST)
 
+gen-cve-feed: ## Generates content from GCS bucket with official cve data.
+	scripts/security/gen-cve-feed.sh
+
```
Hugo can fetch this JSON file itself - see https://gohugo.io/templates/data-templates/

I think I'd prefer that, so we're not modifying the website source during a build.

PushkarJ replied:
Yes good call. It took some work, but got this working without build changes

sftim commented on layouts/shortcodes/issues-security.html:
```diff
@@ -0,0 +1,19 @@
+  </tr>
+  </thead>
+  <tbody>
+  {{ range $issues := getJSON "static/security/official-cve-feed.json" }}
```
This could be a `getJSON` doing a HTTP (well, HTTPS) fetch. Hugo will cache the result.

sftim commented on layouts/shortcodes/issues-security.html:
```diff
@@ -0,0 +1,19 @@
+  {{ range $issues := getJSON "static/security/official-cve-feed.json" }}
+  <tr>
+    <td><a href="{{ .cve_url }}">{{ .cve_id | htmlEscape }}</a></td>
+    <td>{{ .summary | htmlEscape }}</td>
```
(nit) Mark this text as English - it's not localized.
```suggestion
    <td lang="en">{{ .summary | htmlEscape }}</td>
```

PushkarJ commented:
@sftim Thank you for the excellent feedback! I think the [example](https://www.raymondcamden.com/2017/05/18/creating-a-json-feed-for-hugo) you shared on slack, really helped me understand what and how to do this. 

Since with these changes there are no build changes needed. I squashed and rebased the commits so please review this as _almost_ a fresh commit / PR

sftim commented on content/en/docs/reference/issues-security/official-cve-feed.md:
````diff
@@ -0,0 +1,34 @@
+  curl -v https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json
+  ```
+
+{{< official-cve-feed >}}
````
After alpha, it'd be nice to sort the table that this renders.

PushkarJ replied:
Yes agreed. I have added it in future scope here https://github.com/kubernetes/sig-security/issues/1

sftim commented on layouts/_default/official-cve-feed.json:
```diff
@@ -0,0 +1,19 @@
+  "description": "Auto-refreshing official CVE feed for Kubernetes repository",
+  "author": "Kubernetes Community",
+  "items": [
+  {{ range $i, $e := getJSON "https://storage.googleapis.com/fake-test-cve-feed-kep-3203/official-cve-feed.json" }}
```
This doesn't look right for the final piece. Should we mark this PR as a work-in-progress?

sftim replied:
Would `getJSON "URL_GOES_HERE" | jsonify` also work?

PushkarJ replied:
Yes this bucket creation is pending: https://github.com/kubernetes/k8s.io/pull/4009

On the other comment, I am curious what would be the benefit of adding `| jsonify` to the line? 

If I add it it is failing with this error:
`Rebuild failed: failed to render pages: render of "page" failed: "/src/layouts/_default/cve-feed.json:9:115": execute of template failed at <jsonify>: range can’t iterate over `

sftim commented on layouts/shortcodes/official-cve-feed.html:
```diff
@@ -0,0 +1,19 @@
+  </tr>
+  </thead>
+  <tbody>
+  {{ range $issues := getJSON "https://storage.googleapis.com/fake-test-cve-feed-kep-3203/official-cve-feed.json" }}
```
Ideally, we'd make this a site parameter (see the top-level `config.toml`) so that we can change / configure it in one place.

PushkarJ replied:
Added. It does look cleaner

sftim commented on content/en/docs/reference/issues-security/official-cve-feed.md:
```diff
@@ -0,0 +1,34 @@
+outputs:
+  - json
+  - html 
+layout: official-cve-feed
```
I recommend _not_ using the same name for the shortcode and the layout. For example, name the layout `cve-feed` instead.

PushkarJ replied:
Done

sftim commented on content/en/docs/reference/issues-security/official-cve-feed.md:
```diff
@@ -0,0 +1,34 @@
+{{< feature-state for_k8s_version="v1.25" state="alpha" >}}
+
+This is a community maintained list of official CVEs announced by
+Kubernetes Security Response Committee. 
```
```suggestion
the Kubernetes Security Response Committee. See
[Kubernetes Security and Disclosure Information](/docs/reference/issues-security/security/)
for more details.
```

sftim commented on content/en/docs/reference/issues-security/official-cve-feed.md:
```diff
@@ -0,0 +1,34 @@
+This is a community maintained list of official CVEs announced by
+Kubernetes Security Response Committee. 
+
+A programmatically accessible `JSON` Feed can be found [here](./index.json).
```
```suggestion
The Kubernetes project publishes a programmatically accessible
[JSON Feed](/docs/reference/issues-security/official-cve-feed/index.json) of published security issues.
```

sftim commented on content/en/docs/reference/issues-security/official-cve-feed.md:
````diff
@@ -0,0 +1,34 @@
+A programmatically accessible `JSON` Feed can be found [here](./index.json).
+You can access it by executing the following command:
+  ```shell
+  curl -v https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json
````
I'd write
```shell
curl -v https://k8s.io/docs/reference/issues-security/official-cve-feed/index.json
```
Also, don't indent these lines.

PushkarJ replied:
Done

sftim commented on content/en/docs/reference/issues-security/official-cve-feed.md:
```suggestion
The source of truth of this feed is a set of GitHub Issues, filtered by a controlled and
```

sftim reviewed (commented):
This is great!

Some suggestions. But still, this is great!

sftim commented:
/retitle [WIP] [KEP-3203] Fetch and Render CVE JSON feed

sftim commented on content/en/docs/reference/issues-security/issues.md:
```diff
@@ -8,6 +8,7 @@ To report a security issue, please follow the [Kubernetes security disclosure pr
 Work on Kubernetes code and public issues are tracked using [GitHub Issues](https://github.com/kubernetes/kubernetes/issues/).
 
-* [CVE-related issues](https://github.com/kubernetes/kubernetes/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3Aarea%2Fsecurity+in%3Atitle+CVE)
+* [Official List of known CVEs](./../official-cve-feed) announced by [Security Response Committee](https://github.com/kubernetes/committee-security-response)
```
```suggestion
* Official [list of known CVEs](/docs/reference/issues-security/official-cve-feed/)
  (security vulnerabilities) that have been announced by the
  [Security Response Committee](https://github.com/kubernetes/committee-security-response)
```
?

The key change here is to use a site-relative path rather than `./../official-cve-feed`

PushkarJ replied:
I always miss this :( Thanks for catching it

sftim commented on layouts/_default/official-cve-feed.json:
```diff
@@ -0,0 +1,19 @@
+  {{ range $i, $e := getJSON "https://storage.googleapis.com/fake-test-cve-feed-kep-3203/official-cve-feed.json" }}
+    {{ if $i }}, {{ end }}
+      {
+        "{{ T "cve_id" }}": "{{ .cve_id | htmlEscape }}",
```
Another option:
```suggestion
        {{ T "cve_id" | jsonify }}: {{ .cve_id | jsonify }},
```
If that works how I think it would, repeat the same approach for the other lines.

PushkarJ replied:
Yes this works. Updated

PushkarJ commented:
/label tide/merge-method-squash

(@sftim All the working suggestions are incorporated in this commit: https://github.com/kubernetes/website/pull/35228/commits/a3ea562a87bebf1adea2835e250e8c8556365b06)

PushkarJ commented on config.toml:
```diff
@@ -169,6 +169,9 @@ algolia_docsearch = false
 offlineSearch = false
 
+# Official CVE feed bucket URL
+cveFeedBucket = "https://storage.googleapis.com/fake-test-cve-feed-kep-3203/official-cve-feed.json"
```
TODO: This needs an update before merging

PushkarJ commented:
/retitle [KEP-3203] Fetch and Render CVE JSON Feed

(The bucket used is official one and has the recent data)

PushkarJ commented:
/hold cancel

(Since bucket has data already)

sftim commented on layouts/shortcodes/cve-feed.html:
```diff
@@ -0,0 +1,19 @@
+  <tbody>
+  {{ range $issues := getJSON .Site.Params.cveFeedBucket }}
+  <tr>
+    <td><a href="{{ .cve_url }}">{{ .cve_id | htmlEscape }}</a></td>
+    <td>{{ .summary | htmlEscape }}</td>
```
These might need to be:
```suggestion
   <td><a href="{{ .cve_url }}">{{ .cve_id | htmlEscape | safeHTML }}</a></td>
   <td>{{ .summary | htmlEscape | safeHTML }}</td>
```

Look for `&gt;` and `&lt;` in the preview. Marking these escaped strings as safe should fix that (untested).

PushkarJ commented on layouts/shortcodes/cve-feed.html:
```diff
@@ -0,0 +1,19 @@
+  <tbody>
+  {{ range $issues := getJSON .Site.Params.cveFeedBucket }}
+  <tr>
+    <td><a href="{{ .cve_url }}">{{ .cve_id | htmlEscape | safeHTML }}</a></td>
+    <td>{{ .summary | htmlEscape | safeHTML }}</td>
```
@sftim Adding `safeHTML` did the trick here

PushkarJ commented on layouts/_default/cve-feed.json:
```diff
@@ -0,0 +1,24 @@
+        {{ T "cve_json_id" | jsonify }}: {{ .cve_id | jsonify }},
+        {{ T "cve_json_url" | jsonify }}: {{ .issue_url | jsonify }},
+        {{ T "cve_json_external_url" | jsonify }}: {{ .cve_url | jsonify}},
+        {{ T "cve_json_summary" | jsonify }}: {{ replace (.summary | jsonify ) "\\u003e" ">"  }}
```
@sftim but adding it here did not fix things. Turns out it is a known issue: https://github.com/gohugoio/hugo/issues/7229 so I used the solution described in the issue

sftim replied:
I recommend adding a (Hugo) comment to explain why that replace is happening.

PushkarJ replied:
Done. Hugo did not allow me to add comment in JSON file so added it in the corresponding markdown file

sftim commented:
Rendered page LGTM

sftim commented:
Should this PR target the _dev-1.25_ branch?

(I'm not sure - do we want to add the feed with the v1.25 release, or perhaps actually combine the blog announcement for the new feed with the switch-on)

We have options, and several of them are good. I want to be clear on which option we're picking so I can based my approval decision on that context.

PushkarJ commented:
@sftim I think the #35608 blog being published along with the CVE feed being available by v1.25 sounds like a good idea to me. I have switched target branch to `dev-1.25` now.

sftim commented:
> @sftim I think the https://github.com/kubernetes/website/pull/35608 blog being published along with the CVE feed being available by v1.25 sounds like a good idea to me

I'm not sure I understand. Which of these is right:
- the feed goes live on v1.25 release day
- the feed goes live after v1.25 release day, simultaneous with the blog announcement about the new feed going live
- something else!

?

PushkarJ commented:
What I meant was, the feed goes live on v1.25 release day, the feature blog is followed soon after (in next few days). This will allow the feed to have some soak time to iron out any issues we find after going live :)

sftim commented:
> What I meant was, the feed goes live on v1.25 release day, the feature blog is followed soon after (in next few days). This will allow the feed to have some soak time to iron out any issues we find after going live :)

Sounds good to me.

/approve
/remove-label tide/merge-method-squash

sftim commented:
(I'd like somebody else to add the LGTM here)

PushkarJ commented:
/cc @nehaLohia27 @savitharaghunathan 

(Tagging you both for `/lgtm` and/or review comments)

reylejano commented:
/milestone 1.25
/assign @kcmartin

reylejano commented:
/lgtm

katcosgrove commented:
Hi from the Comms team! We have you tracked for a feature blog, but the link provided to us was a KEP, not a placeholder blog PR. Are you still intending to have a feature blog for this release?

sftim commented:
Announcement blog article will be added via https://github.com/kubernetes/website/pull/35608
